### ISSUE TITLE
Use fastjet plugins?

### ISSUE THREAD

**Issue by skluth** [scikit-hep/fastjet#77]
For our studies we would need access to some of the fastjet algorithms available as plugins. Examples are the eesiscone, the jade and the pxcone e+e- jet algorithms.

We can of course do this directly in C++, or with a dedicated C++ wrapper which can be loaded into python via ROOT. 

But it would be great if the fastjet plugins could be made available directly through python fastjet.

**Comment by aryan26roy** [comment]
@skluth , thank you for opening this issue. Fastjet plugins are going to be added to python fastjet in the future, although I cannot tell you when as it will require a lot of time from someone. Maybe @jpivarski knows more about it.

**Comment by jpivarski** [comment]
If what you're asking about is in fj-contrib, we're waiting for fj-contrib to move from SVN to git so that we can make it a git submodule and include it in the build.

If what you're asking about is in FastJet core, the Python module is built with all plugins enabled. I think only the CGAL support was temporarily turned off because of a compilation problem, but that only affects performance at high pile-up, not capabilities. Oh! We also turned off the one and only plugin that required Fortran. Many of the plugins were different jet algorithms—are eesiscone, jade, and pxcone among that set?

These are the compilation arguments we use:

https://github.com/scikit-hep/fastjet/blob/e5aebdc66167400472cd29a36f4af0f2a789a992/setup.py#L62-L70

Instead of listing out all those plugins, we include all of the ones written in C++ (as opposed to Fortran).

**Comment by aryan26roy** [comment]
@jpivarski I think pxcone was the one that required Fortran.

**Comment by jpivarski** [comment]
Ah, then we'd need to get Fortran into CD. Let's hear from @skluth about the exact requirements, first, though.

**Comment by skluth** [comment]
I think the problem for us is how to use a JetDefinition which needs a plugin even if the plugin enabled. We would like to use the various e+e- algorithms like eesiscone and JADE.

**Comment by jpivarski** [comment]
For how to use a JetDefinition, is what you're looking for covered in https://fastjet.readthedocs.io/en/latest/Awkward.html?

For example,

```python
import fastjet
jetdef = fastjet.JetDefinition(fastjet.antikt_algorithm, 0.6)
```

What happens if you put in `ee_genkt_algorithm` or `ee_kt_algorithm`?

I didn't see any of the others you mentioned in that list. Since this is a Python module, we'd have to have the plugins compiled-in to be able to use them.

**Comment by skluth** [comment]
The compiled-in algorithms work for us. The problem is creating a JetDefinition object with a plugin algo. All plugin algos have a enum fastjet::plugin_algorithm. In C++ one has to create a plugin instance of the correct type and create a JetDefinition instance with that. That probably needs dedicated code which detects that a requested algo is a plugin (i.e. not built-in), creates a C++ instance of that plugin, creates a JetDefintion object with the plugin and passes that back.

**Comment by jpivarski** [comment]
@aryan26roy, is this something that can be done with the SWIG objects, or is it a C++ interface that needs to be exposed into Python?

(First question is, do you recognize the procedure @skluth is describing? I don't know how these things are done in C++ so I don't quite understand the procedure. But since you've been working on the C++ side, maybe it makes more sense to you?)

**Comment by aryan26roy** [comment]
@jpivarski I couldn't find any of the algorithms that @skluth mentioned in the import list from swig. While the procedure for making a JetDefinition for a plugin is the same as native algorithms, if those plugins are not exposed through the swig bindings, we couldn't support them. (I can't find the eesiscone algorithms anywhere, maybe it is a part of fj contrib.)

**Comment by skluth** [comment]
Here is the logic I have in a class to handle plugins:

 if( ja == fastjet::plugin_algorithm ) { // ja derived from input jetalgString
    if( jetalgString == "siscone" ) {
      plugin= new fastjet::SISConePlugin( R, 0.75 );
    }
    else if( jetalgString == "eesiscone" ) {
      plugin= new fastjet::SISConeSphericalPlugin( R, 0.75 );
    }
    else if( jetalgString == "jade" ) {
      plugin= new fastjet::JadePlugin();
      recombiner= new EEE0Recombiner();
    }
    else if( jetalgString == "pxcone" ) {
      plugin= new fastjet::PxConePlugin( R, Emin, 0.75, true, 1 );
    }
    else {
      string txt= "TFastJet::TFastJet: wrong plugin: " + jetalgString;
      throw std::runtime_error( txt );
    }
    jetdef= fastjet::JetDefinition( plugin );
}

**Comment by aryan26roy** [comment]
@skluth thanks! I found all the algorithms you mentioned in the docs. All of these plugin algorithms do not seem to be binded through swig @jpivarski.

**Comment by skluth** [comment]
Perhaps it is enough to expose the plugin classes in the swig import list? Then one could build an instance on the python side and create e.g.

```
import fastjet
pluginAlg = fastjet.JadePlugin()
jetdef = fastjet.JetDefinition( pluginAlg )
```

**Comment by skluth** [comment]
About the Fortran-based pxcone algorithm: it is wrapped in C++ in fastjet, so I would think that all we would need for pxcone is building fastjet with pxcone (needs a Fortran compiler i.e. gfortran) and exposing the relevant plugin class fastjet::PxConePlugin( ... ) through the swig magic.

**Comment by aryan26roy** [comment]
@skluth yes, that should be enough. However, the swig bindings are provided by the core fastjet authors, perhaps we can ask them about it @jpivarski.

**Comment by skluth** [comment]
I have a fastjet-3.4.0 build. I could try to experiment with the swig bindings, but I have no experience with this so far, so I would a little introduction into where to look in the fastjet codebase.

**Comment by jpivarski** [comment]
> @skluth yes, that should be enough. However, the swig bindings are provided by the core fastjet authors, perhaps we can ask them about it @jpivarski.

We can ask why this algorithm isn't configured in the same way as the others, or why it can't be done through the SWIG bindings, but if there's a way to do it in C++, then we can also expose that through pybind11.

**Comment by lgray** [comment]
FYI This is probably much much easier now since we have switched the underlying build system of all of fastjet, and these python bindings, to cmake. Now you can `target_link_libraries(myplugin PUBLIC fastjet::fastjetplugins)`
